src/expr.rs

```diff
@@ -115,11 +116,99 @@ impl Rewrite for ast::Expr {
                 };
                 Some(format!("continue{}", id_str))
             }
+            ast::Expr_::ExprClosure(capture, ref fn_decl, ref body) => {
+                rewrite_closure(capture, fn_decl, body, self.span, context, width, offset)
+            }
             _ => context.codemap.span_to_snippet(self.span).ok(),
         }
     }
 }
 
+// This functions is pretty messy because of the wrapping and unwrapping of
+// expressions into and from blocks. See rust issue #27872.
+fn rewrite_closure(capture: ast::CaptureClause,
+                   fn_decl: &ast::FnDecl,
+                   body: &ast::Block,
+                   span: Span,
+                   context: &RewriteContext,
+                   width: usize,
+                   offset: usize)
+                   -> Option<String> {
+    let mover = if capture == ast::CaptureClause::CaptureByValue {
+        "move "
+    } else {
+        ""
+    };
+    let offset = offset + mover.len();
+
+    // 4 = "|| {".len(), which is overconservative when the closure consists of
+    // a single expression.
+    let argument_budget = try_opt!(width.checked_sub(4 + mover.len()));
+    // 1 = |
+    let argument_offset = offset + 1;
+
+    let arg_items = itemize_list(context.codemap,
+                                 fn_decl.inputs.iter(),
+                                 "|",
+                                 |arg| span_lo_for_arg(arg),
+                                 |arg| span_hi_for_arg(arg),
+                                 |arg| rewrite_fn_input(arg),
+                                 span_after(span, "|", context.codemap),
+                                 body.span.lo);
+
+    let fmt = ListFormatting::for_fn(argument_budget, argument_offset);
+    let prefix = format!("{}|{}|", mover, write_list(&arg_items.collect::<Vec<_>>(), &fmt));
+    let block_indent = closure_block_indent(context, offset);
+
+    let body_rewrite = if body.stmts.is_empty() {
+        let expr = body.expr.as_ref().unwrap();
+        // All closure bodies are blocks in the eyes of the AST, but we may not
+        // want to format them as such when they only contain a single
+        // expression.
+        let inner_expr = match expr.node {
+            ast::Expr_::ExprBlock(ref inner_body) if inner_body.stmts.is_empty() => {
+                inner_body.expr.as_ref().unwrap()
+            }
+            _ => expr,
+        };
+
+        // 1 = the separating space between arguments and the body.
+        let extra_offset = extra_offset(&prefix, offset) + 1;
+        let rewrite = inner_expr.rewrite(context, width - extra_offset, offset + extra_offset);
+
+        // Checks if rewrite succeeded and fits on a single line.
+        let accept_rewrite = rewrite.as_ref().map(|result| !result.contains('\n')).unwrap_or(false);
+
+        if accept_rewrite {
+            rewrite
+        } else {
+            if let ast::Expr_::ExprBlock(ref inner_body) = expr.node {
```

Comment:
It'll look at the span yea. I think the long term solution is to address the rust issue.

-----Original Message-----
From: "Nick Cameron" <notifications@github.com>
Sent: ‎19-‎8-‎2015 23:57
To: "nrc/rustfmt" <rustfmt@noreply.github.com>
Cc: "Marcus Klaas de Vries" <mail@marcusklaas.nl>
Subject: Re: [rustfmt] Format closures (#186)

In src/expr.rs:
> +                inner_body.expr.as_ref().unwrap()
> +            }
> +            _ => expr,
> +        };
> +
> +        // 1 = the separating space between arguments and the body.
> +        let extra_offset = extra_offset(&prefix, offset) + 1;
> +        let rewrite = inner_expr.rewrite(context, width - extra_offset, offset + extra_offset);
> +
> +        // Checks if rewrite succeeded and fits on a single line.
> +        let accept_rewrite = rewrite.as_ref().map(|result| !result.contains('\n')).unwrap_or(false);
> +
> +        if accept_rewrite {
> +            rewrite
> +        } else {
> +            if let ast::Expr_::ExprBlock(ref inner_body) = expr.node {
Can you just use a dummy span or will the block code look at the block's span for comments etc? If you can't use a dummy span, then I don't think it is an option
—
Reply to this email directly or view it on GitHub.